shared/src/metabase/shared/models/visualization_settings.cljc

```diff
@@ -0,0 +1,265 @@
+  already has a click action, it will be replaced."
+  {:added "0.40.0"}
+  [settings col-key action]
+  (-> settings
+      with-col-settings
+      (update-in [::column-settings] #(assoc % col-key {::click-behavior action}))))
+
+(defn with-entity-click-action
+  "Creates a click action from a given `from-field-id` Field identifier to the given `to-entity-type` having ID
+  `to-entity-id`. This happens in the normalized form, and hence this should be passed the output of another fn
+  (including, currently, `visualization-settings`). If the given `from-field-id` already has a click action, it will
+  be replaced."
+  {:added "0.40.0"}
+  [settings from-field-id to-entity-type to-entity-id]
+  (with-click-action settings (column-ref-for-id from-field-id) (entity-click-action to-entity-type to-entity-id)))
+
+(def ^:private db-to-normalized-click-action-type
+  {"link"        ::link
+   "crossfilter" ::cross-filter})
+
+(def ^:private normalized-to-db-click-action-type
+  (set/map-invert db-to-normalized-click-action-type))
+
+(def ^:private db-to-normalized-link-type
+  {"question"    ::card
+   "dashboard"   ::dashboard
+   "url"         ::url})
+
+(def ^:private normalized-to-db-link-type
+  (set/map-invert db-to-normalized-link-type))
+
+(def ^:private db->norm-click-behavior-keys
+  {:parameterMapping ::link-parameter-mapping
+   :targetId         ::link-target-id
+   :linkTemplate     ::link-template})
+
+(def ^:private norm->db-click-behavior-keys
+  (set/map-invert db->norm-click-behavior-keys))
+
+(def ^:private db->norm-column-settings-keys
+  {:column_title      ::column-title
+   :date_style        ::date-style
+   :date_abbreviate   ::date-abbreviate
+   :time_style        ::time-style
+   :time_enabled      ::time-enabled
+   :decimals          ::decimals
+   :number_separators ::number-separators
+   :number_style      ::number-style
+   :prefix            ::prefix
+   :suffix            ::suffix
+   :view_as           ::view-as
+   :link_text         ::link-text})
+
+(s/def ::column-title string?)
+(s/def ::date-style #{"M/D/YYYY" "D/M/YYYY" "YYYY/M/D" "MMMM D, YYYY" "D MMMM, YYYY" "dddd, MMMM D, YYYY"})
+(s/def ::date-abbreviate boolean?)
+(s/def ::time-style #{"h:mm A" "k:mm" "h A"})
+(s/def ::time-enabled #{nil "minutes" "seconds" "milliseconds"})
+(s/def ::decimals pos-int?)
+(s/def ::number-separators #(or nil? (and string? (= 2 (count %)))))
+(s/def ::number-style #{"decimal" "percent" "scientific" "currency"})
+(s/def ::prefix string?)
+(s/def ::suffix string?)
+(s/def ::view-as string?)
+(s/def ::link-text string?)
+
+(def ^:private norm->db-column-settings-keys
+  (set/map-invert db->norm-column-settings-keys))
+
+(defn- db-form-entry-to-normalized
+  "Converts a :column_settings DB form to qualified form. Does the opposite of
+  `db-normalized-entry-to-db-form-entry-to-normalized`."
+  [m k v]
+  (case k
+    :click_behavior (assoc m ::click-behavior (-> v
```

Comment:
It's not clear to me from reading the code, are we handling click behavior defined on table columns or just click behavior defined on the top level?

Reply:
As of now, it's **only** column level click actions.  I think I need to see some examples of top level still...